### PR TITLE
Blockstorage command - DumpInBlock

### PR DESCRIPTION
**Pull request by vncoelho**

### CODE REVIEW

**Comment by vncoelho** [comment]
Complex implementation, off course..

**Comment by diskooooo** [comment]
> 50 commits. Complex implementation, off course..

You do know you can squash commits and force push a WIP branch/PR, right? 😆

**Comment by erikzhang** [comment]
If a data is written in block 100, and is modified in block 200. If you export the states of height 150, the data will be missed.

**Comment by vncoelho** [comment]
Yes, @erik, exactly this.

We use this for dumping the state of every block when re-syncing the chain, at Persist.

Then, we can create files only with the states transitions.

https://github.com/NeoResearch/neo-storage-audit

**Comment by erikzhang** [comment]
> We use this for dumping the state of every block when re-syncing the chain, at Persist.

If so, we do not need `StorageItem.Height`, right?

File: StatesDumper/StatesDumper.cs [hunk 1]
```diff
@@ -42,5 +48,34 @@ protected override bool OnMessage(object message)
...
+            JArray array = new JArray(states.Where(p => 
+            {
+                byte[] src = p.Value.ToArray();
+                StorageItem si = new StorageItem();
+                using (MemoryStream stream = new MemoryStream(src)) {
+                  using (BinaryReader reader = new BinaryReader(stream)) {
+                    si.Deserialize(reader);
+                  }
+                }
+                uint h = si.Height;
```
Comment (vncoelho) [inline StatesDumper/StatesDumper.cs]: We need it here, Erik.

**Comment by vncoelho** [comment]
si.Height after serialization.

**Comment by vncoelho** [comment]
Thanks, @erikzhang, great idea and will probably work much faster than the current Filter.
It will probably solve our goal of exporting the storage quick and precise!

The only tradeoff will be the RPC call `getstorageheight` (PR https://github.com/neo-project/neo/pull/424), which is a tool that is good for transparency, exchanges would probably like.
However, we could create a parallel service for that.

**Comment by vncoelho** [comment]
A draft of the dump is being done here, @erikzhang: https://github.com/vncoelho/neo/commithttps://github.com/vncoelho/neo/commit/e4c86ba486e29d93a6e0f1f4e625ed0fbf967430

**Pull request closed without merging**